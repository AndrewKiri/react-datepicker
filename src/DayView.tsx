import * as classNames from 'classnames';
import * as moment from 'moment';
import * as React from 'react';
import TableCell from './TableCell';
import TableMatrixView from './TableMatrixView';

import { getDayMatrix, isDayEqual, isDayRange } from './util/DateUtil';

export interface Props {
  selected?: moment.Moment[];
  startDay?: moment.Moment;
  endDay?: moment.Moment;
  onClick?: (value: string) => void;
  customDayClass?: (date: moment.Moment) => string | string[];
  customDayText?: (date: moment.Moment) => string;
}

interface PrivateProps {
  current: moment.Moment;
}

class DayView extends React.Component<Props & PrivateProps> {
  public getDayClass = (date: string): string => {
    const { current, customDayClass, startDay, endDay } = this.props;
    const currentDate = moment(current).date(parseInt(date, 10));
    let classArr: string[] = [];

    if (!date.trim()) {
      return '';
    }

    if (customDayClass !== undefined) {
      const customClass = customDayClass(currentDate);
      classArr = classArr.concat(typeof customClass === 'string' ? [customClass] : customClass);
    }

    const dayClass = classNames(
      'calendar__day',
      `calendar__day--${currentDate.day()}`,
      customDayClass,
      {
        'calendar__day--end': isDayEqual(currentDate, endDay),
        'calendar__day--range': isDayRange(currentDate, startDay, endDay),
        'calendar__day--selected': this.isSelected(date),
        'calendar__day--start': isDayEqual(currentDate, startDay),
        'calendar__day--today': isDayEqual(currentDate, current),
      }
    );

    return dayClass;
  };

  public getCustomText = (date: string): string => {
    const { current, customDayText, startDay, endDay } = this.props;
    const currentDate = moment(current).date(parseInt(date, 10));
    if (!date.trim()) {
      return '';
    }
    if (!customDayText) {
      return '';
    }

    return customDayText(currentDate);
  };

  public isSelected = (date: string): boolean => {
    const { selected } = this.props;
    if (selected === undefined) {
      return false;
    }
    return selected.some(v => this.isDayEqual(date, v));
  };

  public isDayEqual = (date: string, compareDate?: moment.Moment): boolean => {
    const { current } = this.props;
    const currentDate = moment(current).date(parseInt(date, 10));
    if (!date.trim()) {
      return false;
    }
    if (!compareDate) {
      return false;
    }
    return isDayEqual(compareDate, currentDate);
  };

  public isRange = (date: string): boolean => {
    const { current, startDay, endDay } = this.props;
    const currentDate = moment(current).date(parseInt(date, 10));

    if (!date.trim()) {
      return false;
    }
    if (!startDay || !endDay) {
      return false;
    }

    return isDayRange(currentDate, startDay, endDay);
  };

  public render() {
    const { current, onClick } = this.props;

    const dayMatrix = getDayMatrix(current.year(), current.month());
    const weekdays = moment.weekdaysShort();

    return (
      <TableMatrixView
        headers={weekdays}
        matrix={dayMatrix}
        cell={(date, key) => (
          <TableCell
            className={this.getDayClass(date)}
            subText={this.getCustomText(date)}
            onClick={onClick}
            text={date}
            key={key}
          />
        )}
      />
    );
  }
}

export default DayView;
