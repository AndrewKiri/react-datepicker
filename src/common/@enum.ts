export namespace CalendarEnums {
  export enum ViewMode {
    YEAR,
    MONTH,
    DAY,
  }
}
