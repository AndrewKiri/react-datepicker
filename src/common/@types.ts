export namespace IDatePicker {
  export enum PickerDirection {
    TOP,
    BOTTOM,
  }
  export interface Position {
    left?: string;
    top?: string;
    bottom?: string;
    right?: string;
  }
}
