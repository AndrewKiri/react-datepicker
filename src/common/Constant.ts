export const DatePickerDefaults = {
  dateFormat: 'YYYY-MM-DD',
  locale: 'en',
};
