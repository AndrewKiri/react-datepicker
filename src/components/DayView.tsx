import * as classNames from 'classnames';
import * as moment from 'moment';
import * as React from 'react';
import TableCell from './TableCell';
import TableMatrixView from './TableMatrixView';
import { ifExistCall } from '../utils/FunctionUtil';

import { getDayMatrix, isDayEqual, isDayRange } from '../utils/DateUtil';

export interface Props {
  /** Selected days to show in calendar */
  selected?: moment.Moment[];
  /** Disable days show in calendar */
  disabled?: moment.Moment[];
  /** Start day to show in calendar */
  startDay?: moment.Moment;
  /** End day to show in calendar */
  endDay?: moment.Moment;
  /** Calendar day click Event */
  onClick?: (date: string) => void;
  /** Calendar day Mouseover Event */
  onMouseOver?: (date: moment.Moment) => void;
  /** Custom day class to show in day */
  customDayClass?: (date: moment.Moment) => string | string[];
  /** Custom day text to show in day */
  customDayText?: (date: moment.Moment) => string;
}

interface PrivateProps {
  current: moment.Moment;
  locale: string;
}

class DayView extends React.Component<Props & PrivateProps> {
  public static defaultProps = {
    locale: 'en-ca',
  };

  public getDayClass = (date: string): string => {
    const { current, customDayClass, startDay, endDay, selected, disabled } = this.props;
    const currentDate = moment(current).date(parseInt(date, 10));
    let classArr: string[] = [];

    if (!date.trim()) {
      return '';
    }

    if (customDayClass !== undefined) {
      const customClass = customDayClass(currentDate);
      classArr = classArr.concat(typeof customClass === 'string' ? [customClass] : customClass);
    }

    const dayClass = classNames('calendar__day', `calendar__day--${currentDate.day()}`, classArr, {
      'calendar__day--end': isDayEqual(currentDate, endDay),
      'calendar__day--range': isDayRange(currentDate, startDay, endDay),
      'calendar__day--selected': this.isIncludeDay(date, selected),
      'calendar__day--disabled': this.isIncludeDay(date, disabled),
      'calendar__day--start': isDayEqual(currentDate, startDay),
      'calendar__day--today': isDayEqual(currentDate, moment()),
    });

    return dayClass;
  };

  public getCustomText = (date: string): string => {
    const { current, customDayText } = this.props;
    const currentDate = moment(current).date(parseInt(date, 10));
    if (!date.trim()) {
      return '';
    }
    if (!customDayText) {
      return '';
    }

    return customDayText(currentDate);
  };

  public isIncludeDay = (date: string, dates?: moment.Moment[]): boolean => {
    const { current } = this.props;
    if (dates === undefined) {
      return false;
    }
    return dates.some(v => isDayEqual(moment(current).date(parseInt(date, 10)), v));
  };

  public handleClick = (date: string) => {
    if(this.isIncludeDay(date, this.props.disabled)) return ;
    ifExistCall(this.props.onClick, date);
  };

  public handleMouseOver = (date: string) => {
    const { onMouseOver, current } = this.props;
    ifExistCall(onMouseOver, moment(current).date(parseInt(date, 10)));
  };


  public render() {
    const { current, locale } = this.props;

    const dayMatrix = getDayMatrix(current.year(), current.month());
    const weekdays = moment.localeData(locale).weekdaysShort();

    return (
      <TableMatrixView
        headers={weekdays}
        matrix={dayMatrix}
        cell={(date, key) => (
          <TableCell
            className={this.getDayClass(date)}
            subText={this.getCustomText(date)}
            onClick={this.handleClick}
            onMouseOver={this.handleMouseOver}
            text={date}
            key={key}
          />
        )}
      />
    );
  }
}

export default DayView;
