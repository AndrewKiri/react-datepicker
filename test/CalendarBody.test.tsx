import { mount, shallow, ReactWrapper, ShallowWrapper } from 'enzyme';
import * as moment from 'moment';
import * as React from 'react';
import * as sinon from 'sinon';
import DayView from '../src/components/DayView';
import TestCalendarBody from '../src/components/CalendarBody';
import { CalendarEnums } from '../src/common/@enum';

describe('<CalendarBody />', () => {
  // 20181201
  const mockMoment = moment.unix(1543622400);
  let shallowComponent: ShallowWrapper<React.Component>;
  let mountComponent: ReactWrapper;

  const CalendarBody = (props: any) => <TestCalendarBody current={mockMoment} {...props} />;

  it('should render correctly', () => {
    shallowComponent = shallow(<CalendarBody />);
    expect(shallowComponent).toBeTruthy();
    expect(shallowComponent).toMatchSnapshot();
  });

  describe('prop: viewMode', () => {
    it('should ViewMode.DAY correctly', () => {
      shallowComponent = shallow(
        <TestCalendarBody current={mockMoment} viewMode={CalendarEnums.ViewMode.DAY} />
      );
      expect(shallowComponent).toMatchSnapshot();
      expect(shallowComponent.find(DayView)).toHaveLength(1);
    });

    it('should ViewMode.MONTH correctly', () => {
      mountComponent = mount(<CalendarBody viewMode={CalendarEnums.ViewMode.MONTH} />);
      expect(mountComponent.find('td.calendar__month')).toHaveLength(12);
    });

    it('should ViewMode.YEAR correctly', () => {
      mountComponent = mount(<CalendarBody viewMode={CalendarEnums.ViewMode.YEAR} />);
      expect(mountComponent.find('td.calendar__year')).toHaveLength(9);
    });
  });
});
