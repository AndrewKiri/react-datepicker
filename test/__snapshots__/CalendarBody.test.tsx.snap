// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CalendarBody /> should render correctly 1`] = `
<div
  className="calendar__body"
>
  <DayView
    current={"2018-12-01T00:00:00.000Z"}
    locale="en-ca"
    onClick={[Function]}
    viewMode={2}
  />
</div>
`;
